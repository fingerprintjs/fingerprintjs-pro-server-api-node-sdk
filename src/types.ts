import { components, paths } from './generatedApiTypes'

export enum Region {
  EU = 'EU',
  AP = 'AP',
  Global = 'Global',
}

export enum AuthenticationMode {
  AuthHeader = 'AuthHeader',
  QueryParameter = 'QueryParameter',
}

/**
 * Options for FingerprintJS server API client
 */
export interface Options {
  /**
   * Secret API key
   */
  apiKey: string
  /**
   * Region of the FingerprintJS service server
   */
  region?: Region
  /**
   * Authentication mode
   * Optional, default value is AuthHeader
   */
  authenticationMode?: AuthenticationMode

  /**
   * Optional fetch implementation
   * */
  fetch?: typeof fetch
}

/**
 * More info: https://dev.fingerprintjs.com/docs/server-api#query-parameters
 */
export type VisitorHistoryFilter = paths['/visitors/{visitor_id}']['get']['parameters']['query']

export type TooManyRequestsError = {
  status: 429
  /**
   * How many seconds to wait before retrying
   */
  retryAfter: number
}

/**
 * More info: https://dev.fingerprintjs.com/docs/server-api#response
 */
export type VisitorsResponse = paths['/visitors/{visitor_id}']['get']['responses']['200']['content']['application/json']
export type VisitorsResponse403 =
  paths['/visitors/{visitor_id}']['get']['responses']['403']['content']['application/json'] & {
    status: 403
  }
export type VisitorsResponse429 =
  paths['/visitors/{visitor_id}']['get']['responses']['429']['content']['application/json']

export type DeleteVisit404Response =
  paths['/visitors/{visitor_id}']['delete']['responses']['404']['content']['application/json']

export type DeleteVisit403Response =
  paths['/visitors/{visitor_id}']['delete']['responses']['403']['content']['application/json']

export type DeleteVisit400Response =
  paths['/visitors/{visitor_id}']['delete']['responses']['400']['content']['application/json'] & {
    status: 400
  }

export type CommonResponse429 = components['schemas']['ErrorCommon429Response']

export type VisitorsError = WithResponse<VisitorsResponse403 | VisitorsResponse429>

export type EventResponse = paths['/events/{request_id}']['get']['responses']['200']['content']['application/json']
export type EventResponse403 = paths['/events/{request_id}']['get']['responses']['403']['content']['application/json']
export type EventResponse404 = paths['/events/{request_id}']['get']['responses']['404']['content']['application/json']

type WithResponse<T> = T & {
  response: Response
}

/**
 * More info: https://dev.fingerprintjs.com/docs/webhooks#identification-webhook-object-format
 */
export type VisitWebhook = components['schemas']['WebhookVisit']
