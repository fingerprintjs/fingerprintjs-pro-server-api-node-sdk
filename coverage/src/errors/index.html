
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for src/errors</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../index.html">All files</a> src/errors</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">91.01% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>81/89</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">63.63% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>35/55</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">76.47% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>26/34</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">93.97% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>78/83</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line high'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="apiErrors.ts"><a href="apiErrors.ts.html">apiErrors.ts</a></td>
	<td data-value="90.9" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 90%"></div><div class="cover-empty" style="width: 10%"></div></div>
	</td>
	<td data-value="90.9" class="pct high">90.9%</td>
	<td data-value="11" class="abs high">10/11</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="66.66" class="pct medium">66.66%</td>
	<td data-value="3" class="abs medium">2/3</td>
	<td data-value="90.9" class="pct high">90.9%</td>
	<td data-value="11" class="abs high">10/11</td>
	</tr>

<tr>
	<td class="file high" data-value="commonErrors.ts"><a href="commonErrors.ts.html">commonErrors.ts</a></td>
	<td data-value="87.5" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 87%"></div><div class="cover-empty" style="width: 13%"></div></div>
	</td>
	<td data-value="87.5" class="pct high">87.5%</td>
	<td data-value="8" class="abs high">7/8</td>
	<td data-value="25" class="pct low">25%</td>
	<td data-value="8" class="abs low">2/8</td>
	<td data-value="50" class="pct medium">50%</td>
	<td data-value="2" class="abs medium">1/2</td>
	<td data-value="87.5" class="pct high">87.5%</td>
	<td data-value="8" class="abs high">7/8</td>
	</tr>

<tr>
	<td class="file high" data-value="eventErrors.ts"><a href="eventErrors.ts.html">eventErrors.ts</a></td>
	<td data-value="90.47" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 90%"></div><div class="cover-empty" style="width: 10%"></div></div>
	</td>
	<td data-value="90.47" class="pct high">90.47%</td>
	<td data-value="21" class="abs high">19/21</td>
	<td data-value="66.66" class="pct medium">66.66%</td>
	<td data-value="24" class="abs medium">16/24</td>
	<td data-value="80" class="pct high">80%</td>
	<td data-value="10" class="abs high">8/10</td>
	<td data-value="94.73" class="pct high">94.73%</td>
	<td data-value="19" class="abs high">18/19</td>
	</tr>

<tr>
	<td class="file high" data-value="getRetryAfter.ts"><a href="getRetryAfter.ts.html">getRetryAfter.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	</tr>

<tr>
	<td class="file high" data-value="unsealError.ts"><a href="unsealError.ts.html">unsealError.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="15" class="abs high">15/15</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="5" class="abs high">5/5</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="14" class="abs high">14/14</td>
	</tr>

<tr>
	<td class="file high" data-value="visitErrors.ts"><a href="visitErrors.ts.html">visitErrors.ts</a></td>
	<td data-value="87.09" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 87%"></div><div class="cover-empty" style="width: 13%"></div></div>
	</td>
	<td data-value="87.09" class="pct high">87.09%</td>
	<td data-value="31" class="abs high">27/31</td>
	<td data-value="66.66" class="pct medium">66.66%</td>
	<td data-value="18" class="abs medium">12/18</td>
	<td data-value="69.23" class="pct medium">69.23%</td>
	<td data-value="13" class="abs medium">9/13</td>
	<td data-value="92.85" class="pct high">92.85%</td>
	<td data-value="28" class="abs high">26/28</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at 2024-10-25T09:54:16.754Z
            </div>
        <script src="../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../sorter.js"></script>
        <script src="../../block-navigation.js"></script>
    </body>
</html>
    