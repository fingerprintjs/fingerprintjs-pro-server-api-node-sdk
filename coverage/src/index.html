
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for src</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../prettify.css" />
    <link rel="stylesheet" href="../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../index.html">All files</a> src</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">94.85% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>166/175</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">83.33% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>45/54</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">96.55% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>28/29</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">94.79% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>164/173</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line high'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="index.ts"><a href="index.ts.html">index.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="9" class="abs high">9/9</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="9" class="abs high">9/9</td>
	</tr>

<tr>
	<td class="file high" data-value="responseUtils.ts"><a href="responseUtils.ts.html">responseUtils.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="6" class="abs high">6/6</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="6" class="abs high">6/6</td>
	</tr>

<tr>
	<td class="file high" data-value="sealedResults.ts"><a href="sealedResults.ts.html">sealedResults.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="39" class="abs high">39/39</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="9" class="abs high">9/9</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="6" class="abs high">6/6</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="39" class="abs high">39/39</td>
	</tr>

<tr>
	<td class="file high" data-value="serverApiClient.ts"><a href="serverApiClient.ts.html">serverApiClient.ts</a></td>
	<td data-value="89.55" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 89%"></div><div class="cover-empty" style="width: 11%"></div></div>
	</td>
	<td data-value="89.55" class="pct high">89.55%</td>
	<td data-value="67" class="abs high">60/67</td>
	<td data-value="63.63" class="pct medium">63.63%</td>
	<td data-value="22" class="abs medium">14/22</td>
	<td data-value="90" class="pct high">90%</td>
	<td data-value="10" class="abs high">9/10</td>
	<td data-value="89.55" class="pct high">89.55%</td>
	<td data-value="67" class="abs high">60/67</td>
	</tr>

<tr>
	<td class="file high" data-value="types.ts"><a href="types.ts.html">types.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="7" class="abs high">7/7</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="7" class="abs high">7/7</td>
	</tr>

<tr>
	<td class="file high" data-value="urlUtils.ts"><a href="urlUtils.ts.html">urlUtils.ts</a></td>
	<td data-value="96.96" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 96%"></div><div class="cover-empty" style="width: 4%"></div></div>
	</td>
	<td data-value="96.96" class="pct high">96.96%</td>
	<td data-value="33" class="abs high">32/33</td>
	<td data-value="91.66" class="pct high">91.66%</td>
	<td data-value="12" class="abs high">11/12</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="7" class="abs high">7/7</td>
	<td data-value="96.77" class="pct high">96.77%</td>
	<td data-value="31" class="abs high">30/31</td>
	</tr>

<tr>
	<td class="file medium" data-value="utils.ts"><a href="utils.ts.html">utils.ts</a></td>
	<td data-value="75" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 75%"></div><div class="cover-empty" style="width: 25%"></div></div>
	</td>
	<td data-value="75" class="pct medium">75%</td>
	<td data-value="4" class="abs medium">3/4</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="75" class="pct medium">75%</td>
	<td data-value="4" class="abs medium">3/4</td>
	</tr>

<tr>
	<td class="file high" data-value="webhook.ts"><a href="webhook.ts.html">webhook.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="10" class="abs high">10/10</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="10" class="abs high">10/10</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at 2025-06-04T13:30:03.627Z
            </div>
        <script src="../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../sorter.js"></script>
        <script src="../block-navigation.js"></script>
    </body>
</html>
    