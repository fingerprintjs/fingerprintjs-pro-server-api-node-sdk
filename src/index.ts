export * from './urlUtils'
export * from './serverApiClient'
export * from './types'
export * from './sealedResults'
export * from './errors/unsealError'
export * from './webhook'
export * from './errors/apiErrors'
