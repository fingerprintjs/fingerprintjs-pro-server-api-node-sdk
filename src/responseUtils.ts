import { SdkError } from './errors/apiErrors'
import { toError } from './utils'

export async function copyResponseJson(response: Response) {
  try {
    return await response.clone().json()
  } catch (e) {
    throw new SdkError('Failed to parse JSON response', response, toError(e))
  }
}
