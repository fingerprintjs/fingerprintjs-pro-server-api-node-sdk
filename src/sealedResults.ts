import { createDecipheriv } from 'crypto'
import { inflateRaw } from 'zlib'
import { promisify } from 'util'
import { EventResponse } from './types'
import { UnsealAggregateError, UnsealError } from './errors/unsealError'
import { Buffer } from 'buffer'

const asyncInflateRaw = promisify(inflateRaw)

export enum DecryptionAlgorithm {
  Aes256Gcm = 'aes-256-gcm',
}

export interface DecryptionKey {
  key: Buffer
  algorithm: DecryptionAlgorithm
}

const SEALED_HEADER = Buffer.from([0x9e, 0x85, 0xdc, 0xed])

function isEventResponse(data: unknown): data is EventResponse {
  return Boolean(data && typeof data === 'object' && 'products' in data)
}

/**
 * @private
 * */
export function parseEventsResponse(unsealed: string) {
  const json = JSON.parse(unsealed)

  if (!isEventResponse(json)) {
    throw new Error('Sealed data is not valid events response')
  }

  return json
}

/**
 * Decrypts the sealed response with the provided keys.
 * The SDK will try to decrypt the result with each key until it succeeds.
 * To learn more about sealed results visit: https://dev.fingerprint.com/docs/sealed-client-results
 */
export async function unsealEventsResponse(sealedData: Buffer, decryptionKeys: DecryptionKey[]) {
  const unsealed = await unseal(sealedData, decryptionKeys)

  return parseEventsResponse(unsealed)
}

/**
 * @private
 * */
export async function unseal(sealedData: Buffer, decryptionKeys: DecryptionKey[]) {
  if (sealedData.subarray(0, SEALED_HEADER.length).toString('hex') !== SEALED_HEADER.toString('hex')) {
    throw new Error('Invalid sealed data header')
  }

  const errors = new UnsealAggregateError([])

  for (const decryptionKey of decryptionKeys) {
    switch (decryptionKey.algorithm) {
      case DecryptionAlgorithm.Aes256Gcm:
        try {
          return await unsealAes256Gcm(sealedData, decryptionKey.key)
        } catch (e) {
          errors.addError(new UnsealError(decryptionKey, e as Error))
          continue
        }

      default:
        throw new Error(`Unsupported decryption algorithm: ${decryptionKey.algorithm}`)
    }
  }

  throw errors
}

async function unsealAes256Gcm(sealedData: Buffer, decryptionKey: Buffer) {
  const nonceLength = 12
  const nonce = sealedData.subarray(SEALED_HEADER.length, SEALED_HEADER.length + nonceLength)

  const authTagLength = 16
  const authTag = sealedData.subarray(-authTagLength)

  const ciphertext = sealedData.subarray(SEALED_HEADER.length + nonceLength, -authTagLength)

  const decipher = createDecipheriv('aes-256-gcm', decryptionKey, nonce).setAuthTag(authTag)
  const compressed = Buffer.concat([decipher.update(ciphertext), decipher.final()])

  const payload = await asyncInflateRaw(compressed)

  return payload.toString()
}
