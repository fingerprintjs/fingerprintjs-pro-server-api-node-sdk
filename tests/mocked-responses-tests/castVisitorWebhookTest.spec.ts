import { VisitWebhook } from '../../src/types'
import visitWebhookBody from './mocked-responses-data/external/webhook.json'

describe('[Mocked body] Cast visitor webhook', () => {
  test('with sample request body', async () => {
    const visit = visitWebhookBody as unknown as VisitWebhook

    expect(visit).toMatchSnapshot()
  })
})
