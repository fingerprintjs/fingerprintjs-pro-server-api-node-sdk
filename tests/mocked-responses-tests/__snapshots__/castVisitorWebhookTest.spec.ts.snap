// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Mocked body] Cast visitor webhook with sample request body 1`] = `
Object {
  "bot": Object {
    "result": "bad",
    "type": "selenium",
  },
  "browserDetails": Object {
    "browserFullVersion": "73.0.3683.86",
    "browserMajorVersion": "73",
    "browserName": "Chrome",
    "device": "Other",
    "os": "Mac OS X",
    "osVersion": "10.14.3",
    "userAgent": "(Macintosh; Intel Mac OS X 10_14_3) Chrome/73.0.3683.86",
  },
  "clientReferrer": "https://google.com?search=banking+services",
  "clonedApp": Object {
    "result": false,
  },
  "confidence": Object {
    "score": 0.97,
  },
  "emulator": Object {
    "result": false,
  },
  "factoryReset": Object {
    "time": "1970-01-01T00:00:00Z",
    "timestamp": 0,
  },
  "firstSeenAt": Object {
    "global": "2022-03-16T11:26:45.362Z",
    "subscription": "2022-03-16T11:31:01.101Z",
  },
  "frida": Object {
    "result": false,
  },
  "incognito": false,
  "ip": "216.3.128.12",
  "ipBlocklist": Object {
    "details": Object {
      "attackSource": false,
      "emailSpam": false,
    },
    "result": false,
  },
  "ipInfo": Object {
    "v4": Object {
      "address": "94.142.239.124",
      "asn": Object {
        "asn": "7922",
        "name": "COMCAST-7922",
        "network": "73.136.0.0/13",
      },
      "dataCenter": Object {
        "name": "DediPath",
        "result": true,
      },
      "geolocation": Object {
        "accuracyRadius": 20,
        "city": Object {
          "name": "Prague",
        },
        "continent": Object {
          "code": "EU",
          "name": "Europe",
        },
        "country": Object {
          "code": "CZ",
          "name": "Czechia",
        },
        "latitude": 50.05,
        "longitude": 14.4,
        "postalCode": "150 00",
        "subdivisions": Array [
          Object {
            "isoCode": "10",
            "name": "Hlavni mesto Praha",
          },
        ],
        "timezone": "Europe/Prague",
      },
    },
  },
  "ipLocation": Object {
    "accuracyRadius": 1,
    "city": Object {
      "name": "Bolingbrook",
    },
    "continent": Object {
      "code": "NA",
      "name": "North America",
    },
    "country": Object {
      "code": "US",
      "name": "United States",
    },
    "latitude": 41.12933,
    "longitude": -88.9954,
    "postalCode": "60547",
    "subdivisions": Array [
      Object {
        "isoCode": "IL",
        "name": "Illinois",
      },
    ],
    "timezone": "America/Chicago",
  },
  "jailbroken": Object {
    "result": false,
  },
  "lastSeenAt": Object {
    "global": "2022-03-16T11:28:34.023Z",
    "subscription": null,
  },
  "linkedId": "any-string",
  "privacySettings": Object {
    "result": false,
  },
  "proxy": Object {
    "result": false,
  },
  "rawDeviceAttributes": Object {
    "architecture": Object {
      "value": 127,
    },
    "audio": Object {
      "value": 35.73832903057337,
    },
    "canvas": Object {
      "value": Object {
        "Geometry": "4dce9d6017c3e0c052a77252f29f2b1c",
        "Text": "dd2474a56ff78c1de3e7a07070ba3b7d",
        "Winding": true,
      },
    },
    "colorDepth": Object {
      "value": 30,
    },
    "colorGamut": Object {
      "value": "srgb",
    },
    "contrast": Object {
      "value": 0,
    },
    "cookiesEnabled": Object {
      "value": true,
    },
  },
  "requestId": "Px6VxbRC6WBkA39yeNH3",
  "rootApps": Object {
    "result": false,
  },
  "tag": Object {
    "requestType": "signup",
    "yourCustomId": 45321,
  },
  "tampering": Object {
    "anomalyScore": 0,
    "result": false,
  },
  "time": "2019-10-12T07:20:50.52Z",
  "timestamp": 1554910997788,
  "tor": Object {
    "result": false,
  },
  "url": "https://banking.example.com/signup",
  "userAgent": "(Macintosh; Intel Mac OS X 10_14_3) Chrome/73.0.3683.86",
  "virtualMachine": Object {
    "result": false,
  },
  "visitorFound": true,
  "visitorId": "3HNey93AkBW6CRbxV6xP",
  "vpn": Object {
    "methods": Object {
      "auxiliaryMobile": false,
      "publicVPN": false,
      "timezoneMismatch": false,
    },
    "result": false,
  },
}
`;
